File: src/rapaio/graphics/base/PchPalette.scala
/*
 * Apache License
 * Version 2.0, January 2004
 * http://www.apache.org/licenses/
 *
 *    Copyright 2013 Aurelian Tutuianu
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package rapaio.graphics.base

import java.awt._

/**
 * @author <a href="mailto:padreati@yahoo.com">Aurelian Tutuianu</a>
 */
object PchPalette {
  private val pchs = Array[Drawer](
    new Drawer('o') {
      def draw(g2d: Graphics2D, x: Int, y: Int, size: Double) {
        g2d.drawOval((x - size).toInt, (y - size).toInt, (size * 2 + 1).toInt, (size * 2 + 1).toInt)
      }
    },
    new Drawer('p') {
      def draw(g2d: Graphics2D, x: Int, y: Int, size: Double) {
        g2d.fillOval((x - size).toInt, (y - size).toInt, (size * 2 + 1).toInt, (size * 2 + 1).toInt)
      }
    },
    new Drawer('s') {
      override def draw(g2d: Graphics2D, x: Int, y: Int, size: Double): Unit = {
        g2d.fillRect((x - size).toInt, (y - size).toInt, (size * 2 + 1).toInt, (size * 2 + 1).toInt)
      }
    },
    new Drawer('+') {
      override def draw(g2d: Graphics2D, x: Int, y: Int, size: Double): Unit = {
        val sz = size.toInt
        g2d.drawLine(x, y - sz - 1, x, y + sz + 1)
        g2d.drawLine(x - sz - 1, y, x + sz + 1, y)
      }
    },
    new Drawer('x') {
      override def draw(g2d: Graphics2D, x: Int, y: Int, size: Double): Unit = {
        val sz = size.toInt
        g2d.drawLine(x - sz, y - sz, x + sz, y + sz)
        g2d.drawLine(x + sz, y - sz, x - sz, y + sz)
      }
    }
  )

  def draw(g2d: Graphics2D, x: Int, y: Int, size: Double, pch: Int) {
    if (pch < 0) draw(g2d, x, y, size, 0)
    if (pch >= pchs.size) draw(g2d, x, y, size, pch % pchs.size)
    pchs(pch).draw(g2d, x, y, size)
  }

  def apply(ch: Char): Int = {
    var result = 0
    for (i <- 0 until pchs.length) {
      if (pchs(i).pattern.equals(ch)) result = i
    }
    result
  }
}

abstract class Drawer(charPattern: Char) {
  def draw(g2d: Graphics2D, x: Int, y: Int, size: Double)

  def pattern: Char = charPattern
}

File: rapaio-core/src/rapaio/graphics/opt/PchPalette.java
/*
 * Apache License
 * Version 2.0, January 2004
 * http://www.apache.org/licenses/
 *
 *    Copyright 2013 Aurelian Tutuianu
 *    Copyright 2014 Aurelian Tutuianu
 *    Copyright 2015 Aurelian Tutuianu
 *    Copyright 2016 Aurelian Tutuianu
 *    Copyright 2017 Aurelian Tutuianu
 *    Copyright 2018 Aurelian Tutuianu
 *    Copyright 2019 Aurelian Tutuianu
 *    Copyright 2020 Aurelian Tutuianu
 *    Copyright 2021 Aurelian Tutuianu
 *    Copyright 2022 Aurelian Tutuianu
 *    Copyright 2023 Aurelian Tutuianu
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 *
 */

package rapaio.graphics.opt;

import static java.lang.StrictMath.sqrt;

import java.awt.BasicStroke;
import java.awt.Color;
import java.awt.Graphics2D;
import java.awt.geom.Ellipse2D;
import java.awt.geom.Line2D;
import java.awt.geom.Path2D;
import java.io.Serializable;
import java.util.ArrayList;

/**
 * @author <a href="mailto:padreati@yahoo.com">Aurelian Tutuianu</a>
 */
public enum PchPalette implements Serializable {

    STANDARD(new StandardPchPalette());

    private final Mapping mapping;

    PchPalette(Mapping mapping) {
        this.mapping = mapping;
    }

    public void draw(Graphics2D g2d, double x, double y, double size, int pch, float lwd, Color color, Color fill) {
        mapping.draw(g2d, x, y, size, pch, lwd, color, fill);
    }

}

interface Mapping {
    void draw(Graphics2D g2d, double x, double y, double size, int pch, float lwd, Color color, Color fill);
}

final class StandardPchPalette implements Mapping {

    private final java.util.List<Drawer> pchs;

    public StandardPchPalette() {
        pchs = new ArrayList<>();

        pchs.add(Drawer.CIRCLE_WIRE);
        pchs.add(Drawer.CIRCLE_FILL);
        pchs.add(Drawer.CIRCLE_FULL);
        pchs.add(Drawer.CROSS_WIRE);
        pchs.add(Drawer.TRIANGLE_WIRE);
        pchs.add(Drawer.TRIANGLE_FILL);
        pchs.add(Drawer.TRIANGLE_FULL);
        pchs.add(Drawer.SQUARE_WIRE);
        pchs.add(Drawer.SQUARE_FILL);
        pchs.add(Drawer.SQUARE_FULL);
    }

    @Override
    public void draw(Graphics2D g2d, double x, double y, double size, int pch, float lwd, Color color, Color fill) {
        if (pch < 0) {
            pch = 0;
        }
        if (pch >= pchs.size()) {
            pch %= pchs.size();
        }
        pchs.get(pch).draw(g2d, x, y, size, lwd, color, fill);
    }
}

enum Drawer {

    CIRCLE_WIRE() {
        @Override
        void draw(Graphics2D g2d, double x, double y, double sz, float lwd, Color color, Color fill) {
            g2d.setStroke(new BasicStroke(lwd));
            g2d.setColor(color);
            g2d.draw(new Ellipse2D.Double(x - sz, y - sz, sz * 2, sz * 2));
        }
    },
    CIRCLE_FILL() {
        @Override
        void draw(Graphics2D g2d, double x, double y, double sz, float lwd, Color color, Color fill) {
            g2d.setStroke(new BasicStroke(1f));
            g2d.setColor(fill);
            g2d.fill(new Ellipse2D.Double(x - sz, y - sz, sz * 2, sz * 2));
        }
    },
    CIRCLE_FULL() {
        @Override
        void draw(Graphics2D g2d, double x, double y, double sz, float lwd, Color color, Color fill) {
            g2d.setColor(fill);
            g2d.setStroke(new BasicStroke(1f));
            g2d.fill(new Ellipse2D.Double(x - sz, y - sz, sz * 2, sz * 2));

            g2d.setColor(color);
            g2d.setStroke(new BasicStroke(lwd));
            g2d.draw(new Ellipse2D.Double(x - sz, y - sz, sz * 2, sz * 2));
        }
    },
    CROSS_WIRE() {
        @Override
        void draw(Graphics2D g2d, double x, double y, double sz, float lwd, Color color, Color fill) {
            g2d.setStroke(new BasicStroke(lwd));
            g2d.setColor(color);
            g2d.draw(new Line2D.Double(x - sz, y - sz, x + sz, y + sz));
            g2d.fill(new Line2D.Double(x - sz, y - sz, x + sz, y + sz));
            g2d.draw(new Line2D.Double(x + sz, y - sz, x - sz, y + sz));
            g2d.fill(new Line2D.Double(x + sz, y - sz, x - sz, y + sz));
        }
    },
    TRIANGLE_WIRE() {
        @Override
        void draw(Graphics2D g2d, double x, double y, double sz, float lwd, Color color, Color fill) {
            sz = sz * 1.2;
            Path2D.Double path = new Path2D.Double();
            path.moveTo(x, y - sz);
            path.lineTo(x + sz * sqrt(3) / 2, y + sz * 2 / 3);
            path.lineTo(x - sz * sqrt(3) / 2, y + sz * 2 / 3);
            path.lineTo(x, y - sz);

            g2d.setStroke(new BasicStroke(lwd));
            g2d.setColor(color);
            g2d.draw(path);
        }
    },
    TRIANGLE_FILL() {
        @Override
        void draw(Graphics2D g2d, double x, double y, double sz, float lwd, Color color, Color fill) {
            sz = sz * 1.2;
            Path2D.Double path = new Path2D.Double();
            path.moveTo(x, y - sz);

            path.lineTo(x + sz * sqrt(3) / 2, y + sz * 2 / 3);
            path.lineTo(x - sz * sqrt(3) / 2, y + sz * 2 / 3);
            path.lineTo(x, y - sz);

            g2d.setStroke(new BasicStroke(1f));
            g2d.setColor(fill);
            g2d.fill(path);
        }
    },
    TRIANGLE_FULL() {
        @Override
        void draw(Graphics2D g2d, double x, double y, double sz, float lwd, Color color, Color fill) {
            sz = sz * 1.2;
            Path2D.Double path = new Path2D.Double();
            path.moveTo(x, y - sz);
            path.lineTo(x + sz * sqrt(3) / 2, y + sz * 2 / 3);
            path.lineTo(x - sz * sqrt(3) / 2, y + sz * 2 / 3);
            path.lineTo(x, y - sz);

            g2d.setStroke(new BasicStroke(1f));
            g2d.setColor(fill);
            g2d.fill(path);

            g2d.setStroke(new BasicStroke(lwd));
            g2d.setColor(color);
            g2d.draw(path);
        }
    },
    SQUARE_WIRE() {
        @Override
        void draw(Graphics2D g2d, double x, double y, double sz, float lwd, Color color, Color fill) {
            Path2D.Double path = new Path2D.Double();
            double d = sz;
            path.moveTo(x - d, y - d);
            path.lineTo(x + d, y - d);
            path.lineTo(x + d, y + d);
            path.lineTo(x - d, y + d);
            path.lineTo(x - d, y - d);

            g2d.setStroke(new BasicStroke(lwd));
            g2d.setColor(color);
            g2d.draw(path);
        }
    },
    SQUARE_FILL() {
        @Override
        void draw(Graphics2D g2d, double x, double y, double sz, float lwd, Color color, Color fill) {
            Path2D.Double path = new Path2D.Double();
            double d = sz;
            path.moveTo(x - d, y - d);
            path.lineTo(x + d, y - d);
            path.lineTo(x + d, y + d);
            path.lineTo(x - d, y + d);
            path.lineTo(x - d, y - d);

            g2d.setStroke(new BasicStroke(1f));
            g2d.setColor(fill);
            g2d.fill(path);
        }
    },
    SQUARE_FULL() {
        @Override
        void draw(Graphics2D g2d, double x, double y, double sz, float lwd, Color color, Color fill) {
            Path2D.Double path = new Path2D.Double();
            path.moveTo(x - sz, y - sz);
            path.lineTo(x + sz, y - sz);
            path.lineTo(x + sz, y + sz);
            path.lineTo(x - sz, y + sz);
            path.lineTo(x - sz, y - sz);

            g2d.setStroke(new BasicStroke(1f));
            g2d.setColor(fill);
            g2d.fill(path);

            g2d.setStroke(new BasicStroke(lwd));
            g2d.setColor(color);
            g2d.draw(path);
        }
    };

    abstract void draw(Graphics2D g2d, double x, double y, double sz, float lwd, Color color, Color fill);
}
